File: src/main/java/pl/edu/pw/elka/mtoporow/cevolver/lib/model/microstrip/Microstrip.scala
package pl.edu.pw.elka.mtoporow.cevolver.lib.model.microstrip

import org.apache.commons.math3.complex.Complex
import pl.edu.pw.elka.mtoporow.cevolver.lib.model.matrix.{SMatrix, TMatrix}
import pl.edu.pw.elka.mtoporow.cevolver.lib.model.util.PhysicalConstants

/**
 * Mikropasek
 * Data utworzenia: 14.06.15, 18:53
 *
 * @param w wysokość paska
 * @param l długość paska
 * @param t wysokość paska
 * @param h wysokość dielektrycznego podłoża
 * @param epsr względna przenikalność elektryczna podłoża
 *
 * @author Michał Toporowski
 */
class Microstrip(
                  val w: Double,
                  val l: Double,
                  val t: Double,
                  val h: Double,
                  val epsr: Double) {

  private val wph = w / h

  /**
   * Względna skuteczna przenikalność elektryczna (eps_ef)
   */
  private val epsef = {
    // względna skuteczna przenikalność elektr. (Korszeń, s. 18)
    if (wph < 1.0) {
      (epsr + 1) / 2 + ((epsr - 1) / 2) * (1.0 / math.sqrt(1.0 + 12.0 * h / w) + 0.041 * (1.0 - wph) * (1.0 - wph))
    } else {
      (epsr + 1) / 2 + ((epsr - 1) / 2) * (1.0 / math.sqrt(1.0 + 12.0 * h / w))
    }
  }

  /**
   * Oblicza impedancję charakterystyczną (Z0)
   */
  def charImpedance(): Double = {
    // obliczenie (Korszeń s. 18) - różne wzory dla w / h < 1 i w / h >= 1
    // uwaga: do rozważenia - dla dużej precyzji można zastąpić szerokość szerokością skuteczną (w_s)
    if (wph < 1.0) {
      (60 / math.sqrt(epsef)) * math.log(8 * h / w + w / (4 * h))
    } else {
      120 * math.Pi / (math.sqrt(epsef) * (wph + 1.393 + (2.0 / 3.0) * math.log(wph + 1.444)))
    }
  }

  /**
   * Oblicza długość elektryczną (theta)
   * @param f częstotliwość fali
   * @return
   */
  private def electricalLength(f: Double): Double = {
    // prędkość fazowa fali - Korszeń, s. 19
    def vphi = PhysicalConstants.LIGHT_SPEED_MPS / math.sqrt(epsef)
    // długość elektryczna
    2 * math.Pi * l * f / vphi
  }

  /**
   * Oblicza macierz S mikropaska
   *
   * @param f częstotliwość fali
   * @param z01 impedancja charakterystyczna prowadnic mikrofalowych podłączonych do wrót linii Z01
   * @return
   */
  def sMatrix(f: Double, z01: Complex): SMatrix = {
    // wyznaczenie macierzy S (Korszeń s. 20)
    // s11 = s22 = j(r^2 - 1)sin(theta) / 2rcos(theta) + j(r^2 + 1)sin(theta), gdzie r = z0/z01
    val theta = electricalLength(f)
    val z0 = Complex.valueOf(charImpedance())
    val r = z0.divide(z01)
    val r2 = r.multiply(r)
    val denominator = r.multiply(2.0).multiply(math.cos(theta)).add(Complex.I.multiply(r2.add(1)).multiply(math.sin(theta)))
    val s11 = Complex.I.multiply(r2.subtract(1)).multiply(math.sin(theta))
      .divide(denominator)
    val s22 = s11
    // s12 = s21 = 2r / 2rcos(theta) + j(r^2 + 1)sin(theta)
    val s12 = r.multiply(2.0).divide(denominator)
    val s21 = s12
    new SMatrix(s11, s12, s21, s22)
  }

  /**
   * Oblicza macierz T mikropaska
   *
   * @param f częstotliwość fali
   * @param z01 impedancja charakterystyczna prowadnic mikrofalowych podłączonych do wrót linii Z01
   * @return
   */
  def tMatrix(f: Double, z01: Complex): TMatrix = {
    // wyznaczenie macierzy T (Korszeń s. 20)
    val theta = electricalLength(f)
    val z0 = Complex.valueOf(charImpedance())
    val r = z0.divide(z01)
    // t11 = cos(theta) - j * 1/2 (r + 1/r) sin(theta)
    // t22 = cos(theta) + j * 1/2 (r + 1/r) sin(theta)
    val costheta = math.cos(theta)
    val sintheta = math.sin(theta)
    val commonSubstrate = Complex.I.multiply(0.5).multiply(r.add(Complex.ONE.divide(r))).multiply(sintheta)
    val t11 = commonSubstrate.add(costheta)
    val t22 = commonSubstrate.subtract(costheta)
    // t12 = -t21 =  j * 1/2 (r - 1/r) sin(theta)
    val t12 = Complex.I.multiply(0.5).multiply(r.add(Complex.ONE.divide(r))).multiply(sintheta)
    val t21 = t12.negate()
    new TMatrix(t11, t12, t21, t22)
  }


}

File: src/main/java/pl/edu/pw/elka/mtoporow/cevolver/lib/model/microstrip/Microstrip.java
package pl.edu.pw.elka.mtoporow.cevolver.lib.model.microstrip;

import org.apache.commons.math3.complex.Complex;
import pl.edu.pw.elka.mtoporow.cevolver.lib.model.matrix.TMatrix;
import pl.edu.pw.elka.mtoporow.cevolver.lib.util.maths.PhysicalConstants;

/**
 * Alternatywny model linii mikropaskowej
 * Data utworzenia: 29.01.16, 15:57
 *
 * @author Michał Toporowski
 */
public class Microstrip {
    private final double w;
    private final double t;
    private final double l;
    private final double h;
    private final double epsr;
    private double z0;
    private double eef;

    /**
     * @param w    wysokość paska (m)
     * @param l    długość paska (m)
     * @param t    wysokość paska (m)
     * @param h    wysokość dielektrycznego podłoża (m)
     * @param epsr względna przenikalność elektryczna podłoża
     */
    public Microstrip(double w, double t, double l, double h, double epsr) {
        this.w = w;
        this.t = t;
        this.l = l;
        this.h = h;
        this.epsr = epsr;
        this.calculate();
    }

    /**
     * Liczy eef oraz Z0
     */
    private void calculate() {
        double wph = w / h;
        if (wph < 1) {
            eef = (epsr + 1.) / 2. + ((epsr - 1.) / 2.) * (1. / Math.sqrt(1. + 12. / wph) + 0.041 * (1 - wph) * (1 - wph));
            z0 = (60. / Math.sqrt(eef)) * Math.log(8. / wph + 0.25 * wph);
        } else {
            eef = (epsr + 1.) / 2. + ((epsr - 1.) / 2.) * (1. / Math.sqrt(1. + 12. / wph));
            z0 = 120. * Math.PI / (Math.sqrt(eef) * (wph + 1.393 + (2. / 3.) * Math.log(wph + 1.444)));
        }
    }

    /**
     * Liczy długość elektryczną
     *
     * @param f częstotliwość (Hz)
     * @return długość elektryczna
     */
    public double getELen(double f) {
        double vphi = PhysicalConstants.LIGHT_SPEED_MPS() / Math.sqrt(eef);
        return 2. * Math.PI * l * f / vphi;
    }

    /**
     * Liczy macierz T dla tego mikropaska
     *
     * @param z01  impedancja (om)
     * @param freq częstotliwość (Hz)
     * @return Macierz T
     */
    public TMatrix getTMatrix(Complex z01, double freq) {
        double theta = getELen(freq);
        Complex r = Complex.valueOf(z0).divide(z01);
        Complex rInv = Complex.ONE.divide(r);
        Complex commonSubstrate = Complex.I.multiply(0.5).multiply(r.add(rInv)).multiply(Math.sin(theta));
        Complex t11 = Complex.valueOf(Math.cos(theta)).subtract(commonSubstrate);
        Complex t22 = Complex.valueOf(Math.cos(theta)).add(commonSubstrate);
        Complex t12 = Complex.I.multiply(0.5).multiply(r.subtract(rInv)).multiply(Math.sin(theta));
        Complex t21 = t12.negate();
        return new TMatrix(t11, t12, t21, t22);
    }


    /**
     * Zwraca impedancję charakterystyczną tego mikropaska
     *
     * @return impedancja
     */
    public double getZ0() {
        return z0;
    }


}
